File: frontend/lib/agentCaps.ts
"use client";

// ── Active agent attach-tokens ───────────────────────────────────────────────
//
// GET/DELETE /api/agent-caps[/:id] — every live agent attach-credential
// across every session you own, with a revoke action. Server-scoped to
// sessions you actually own (crates/db/src/tokens.rs::list_root_caps_for_owner/
// revoke_owned_root_cap) — the id here can never revoke a cap in a session
// you don't own no matter what's passed.

import { API_BASE } from "./env";
import { authFetch } from "./auth";

export type AgentCap = {
  cap_id: string;
  session_id: string;
  session_name: string;
  actor_id: string;
  created_at: string;
  expires_at: string;
  used_at: string | null;
};

export async function listAgentCaps(): Promise<AgentCap[]> {
  const res = await authFetch(`${API_BASE}/agent-caps`);
  if (!res.ok) throw new Error(await res.text() || `HTTP ${res.status}`);
  return res.json();
}

export async function revokeAgentCap(capId: string): Promise<void> {
  const res = await authFetch(`${API_BASE}/agent-caps/${encodeURIComponent(capId)}`, {
    method: "DELETE",
  });
  if (!res.ok) throw new Error(await res.text() || `HTTP ${res.status}`);
}

File: crates/server/src/routes/agent_caps.rs
//! `GET /api/agent-caps` / `DELETE /api/agent-caps/:id` — the Settings
//! "active agent sessions" view: every live agent attach-credential across
//! every session the signed-in actor owns, with a revoke action.
//!
//! Mirrors activity.rs's shape (cross-session, sp_token-scoped to the
//! caller's own identity, no per-row membership check needed because the
//! underlying query is already scoped to sessions the caller owns).

use std::sync::Arc;

use axum::{
    extract::{Path, State},
    http::{HeaderMap, StatusCode},
    response::IntoResponse,
    routing::get,
    Json, Router,
};
use serde_json::json;

use crate::state::AppState;

pub fn router() -> Router<Arc<AppState>> {
    Router::new()
        .route("/", get(list_mine))
        .route("/:id", axum::routing::delete(revoke))
}

async fn list_mine(
    headers:      HeaderMap,
    State(state): State<Arc<AppState>>,
) -> impl IntoResponse {
    let actor_id = match crate::auth::require_sp_auth(&state.db, &headers).await {
        Ok(id)   => id,
        Err(res) => return res,
    };
    match db::tokens::list_root_caps_for_owner(&state.db, &actor_id).await {
        Ok(rows) => {
            let out: Vec<serde_json::Value> = rows.iter().map(|r| json!({
                "cap_id":       r.cap_id,
                "session_id":   r.session_id,
                "session_name": r.session_name,
                "actor_id":     r.actor_id,
                "created_at":   r.created_at,
                "expires_at":   r.expires_at,
                "used_at":      r.used_at,
            })).collect();
            Json(out).into_response()
        }
        Err(e) => (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()).into_response(),
    }
}

async fn revoke(
    headers:      HeaderMap,
    Path(id):     Path<String>,
    State(state): State<Arc<AppState>>,
) -> impl IntoResponse {
    let actor_id = match crate::auth::require_sp_auth(&state.db, &headers).await {
        Ok(id)   => id,
        Err(res) => return res,
    };
    match db::tokens::revoke_owned_root_cap(&state.db, &id, &actor_id).await {
        Ok(Some(_revoked_ids)) => StatusCode::NO_CONTENT.into_response(),
        Ok(None)               => StatusCode::NOT_FOUND.into_response(),
        Err(e)                 => (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()).into_response(),
    }
}
